src/shadow.rs

```diff
@@ -0,0 +1,224 @@
+        lint_array!(SHADOW_SAME, SHADOW_REUSE, SHADOW_FOREIGN)
+    }
+
+    fn check_fn(&mut self, cx: &Context, _: FnKind, decl: &FnDecl,
+            block: &Block, _: Span, _: NodeId) {
+        if in_external_macro(cx, block.span) { return; }
+        check_fn(cx, decl, block);
+    }
+}
+
+fn check_fn(cx: &Context, decl: &FnDecl, block: &Block) {
+    let mut bindings = Vec::new();
+    for arg in &decl.inputs {
+        if let PatIdent(_, ident, _) = arg.pat.node {
+            bindings.push(ident.node.name)
+        }
+    }
+    check_block(cx, block, &mut bindings);
+}
+
+fn named(pat: &Pat) -> Option<Name> {
+    if let PatIdent(_, ident, _) = pat.node {
+       Some(ident.node.name)
+    } else { None }
+}
+
+fn add(bindings: &mut Vec<Name>, pat: &Pat) {
+    named(pat).map(|name| bindings.push(name));
+}
+
+fn check_block(cx: &Context, block: &Block, bindings: &mut Vec<Name>) {
+    let len = bindings.len();
+    for stmt in &block.stmts {
+        match stmt.node {
+            StmtDecl(ref decl, _) => check_decl(cx, decl, bindings),
+            StmtExpr(ref e, _) | StmtSemi(ref e, _) =>
+                check_expr(cx, e, bindings),
+            _ => ()
+        }
+    }
+    if let Some(ref o) = block.expr { check_expr(cx, o, bindings); }
+    bindings.truncate(len);
+}
+
+fn check_decl(cx: &Context, decl: &Decl, bindings: &mut Vec<Name>) {
+    if in_external_macro(cx, decl.span) { return; }
+    if let DeclLocal(ref local) = decl.node {
+        let Local{ ref pat, ref ty, ref init, id: _, span: _ } = **local;
+        if let &Some(ref t) = ty { check_ty(cx, t, bindings); }
+        named(pat).map(|name| if bindings.contains(&name) {
+            if let &Some(ref o) = init {
+                if in_external_macro(cx, o.span) { return; }
+                check_expr(cx, o, bindings);
+                bindings.push(name);
+                lint_shadow(cx, name, decl.span, pat.span, o);
+            }
+        });
+        add(bindings, pat);
+        if let &Some(ref o) = init {
+            check_expr(cx, o, bindings)
+        }
+    }
+}
+
+fn lint_shadow(cx: &Context, name: Name, span: Span, lspan: Span, init: &Expr) {
+    if is_self_shadow(name, init) {
+        span_lint(cx, SHADOW_SAME, span, &format!(
+            "{} is shadowed by itself in {}",
+            snippet(cx, lspan, "_"),
+            snippet(cx, init.span, "..")));
+    } else {
+        if contains_self(name, init) {
+            span_lint(cx, SHADOW_REUSE, span, &format!(
+                "{} is shadowed by {} which reuses the original value",
+                snippet(cx, lspan, "_"),
+                snippet(cx, init.span, "..")));
+        } else {
+            span_lint(cx, SHADOW_FOREIGN, span, &format!(
+                "{} is shadowed by {} in this declaration",
+                snippet(cx, lspan, "_"),
+                snippet(cx, init.span, "..")));
+        }
+    }
+}
+
+fn check_expr(cx: &Context, expr: &Expr, bindings: &mut Vec<Name>) {
```

Comment:
This won't handle branches (I was wondering why this code was much shorter than expected :P ). Two bindings with the same name will lint even if in disjoint branches.

We need a stateful stack-y visitor (not necessary to use syntax::visit) here. humpty_dumpty does this, but its much more complicated -- keeping track of vec.len on the visitor call stack and popping excess should be enough.